testctrl/svc/orch/object.go

```diff
@@ -0,0 +1,225 @@
+
+// ErrorContainerCrashed indicates that an object's docker container is in the waiting state, but
+// a crash has been detected.
+var ErrorContainerCrashed error
+
+// ErrorPodFailed indicates that kubernetes marked an object's pod as failed.
+var ErrorPodFailed error
+
+// init, although normally discouraged, is only used to initialize the error variables, since they
+// require the Errorf function from the fmt package.
+func init() {
+	ErrorContainerTerminated = fmt.Errorf("container terminated")
+	ErrorContainerTerminating = fmt.Errorf("container terminating")
+	ErrorContainerCrashed = fmt.Errorf("container crashed")
+	ErrorPodFailed = fmt.Errorf("pod failed")
+}
+
+// Object is a Component coupled with its status, health, and kubernetes pod information. It is
+// designed to be used internally. It should be instantiated with the NewObjects function. All
+// methods are thread-safe.
+type Object struct {
+	component *types.Component
+	podStatus v1.PodStatus
+	health    Health
+	mux       sync.Mutex
+	err       error
+}
+
+// NewObjects is a constructor for an object that takes components as variadic arguments. For each
+// component, it will return exactly one Object instance in the returned slice.
+func NewObjects(cs ...*types.Component) []*Object {
+	var objects []*Object
+
+	for _, component := range cs {
+		objects = append(objects, &Object{
+			component: component,
+		})
+	}
+
+	return objects
+}
+
+// Name provides convenient access to the component name.
+func (o *Object) Name() string {
+	return o.component.Name()
+}
+
+// Component is the component instance that the object wraps.
+func (o *Object) Component() *types.Component {
+	return o.component
+}
+
+// Update modifies a component's health and status by looking for errors in a kubernetes PodStatus.
+func (o *Object) Update(status v1.PodStatus) {
+	var err error
+	phase := status.Phase
+	if phase == v1.PodFailed {
+		err = ErrorPodFailed
+	}
+
+	var cstate containerState
+	if cstatuses := status.ContainerStatuses; len(cstatuses) > 0 {
+		cstatus := &status.ContainerStatuses[0]
+		if wcstate := cstatus.State.Waiting; wcstate != nil {
+			if strings.Compare("CrashLoopBackOff", wcstate.Reason) == 0 {
+				cstate = containerStateCrashWaiting
+				err = ErrorContainerCrashed
+			} else {
+				cstate = containerStateWaiting
+			}
+		} else if cstatus.State.Terminated != nil {
+			cstate = containerStateTerminated
+			err = ErrorContainerTerminated
+		} else if cstatus.LastTerminationState.Terminated != nil {
+			cstate = containerStateTerminating
+			err = ErrorContainerTerminating
+		} else if cstatus.State.Running != nil {
+			cstate = containerStateRunning
+		} else {
+			cstate = containerStateUnknown
+		}
+	}
+
+	o.mux.Lock()
+	defer o.mux.Unlock()
+
+	psm, ok := phaseStateMap[phase]
+	if !ok {
+		o.health = Unknown
+	}
+
+	health, ok := psm[cstate]
+	if ok {
+		o.health = health
+	} else {
+		o.health = Unknown
+	}
+
+	o.podStatus = status
+	o.err = err
```

Comment:
TODO( @codeblooded ): Need to add error tests and prevent `err` from being set with `v1.PodSucceeded`